src/v/storage/index_state.cc

```diff
@@ -29,6 +30,152 @@
+}
+
+std::optional<int>
+index_columns::position_upper_bound(uint64_t needle) const noexcept {
+    auto it = std::upper_bound(
+      std::begin(_position_index),
+      std::end(_position_index),
+      needle,
+      std::less<uint32_t>());
+    if (it == std::end(_position_index)) {
+        return std::nullopt;
+    }
+    return std::distance(std::begin(_position_index), it);
+}
+
+std::optional<int>
+index_columns::time_lower_bound(uint32_t needle) const noexcept {
+    auto it = std::lower_bound(
+      std::begin(_relative_time_index),
+      std::end(_relative_time_index),
+      needle,
+      std::less<uint32_t>());
+    if (it == std::end(_relative_time_index)) {
+        return std::nullopt;
+    }
+    return std::distance(std::begin(_relative_time_index), it);
+}
+
+bool index_columns::try_reset_relative_time_index(uint32_t t) {
+    if (_relative_time_index.size() != 1) {
+        return false;
+    }
+    _relative_time_index[0] = t;
+    return true;
+}
+
+chunked_vector<uint32_t>
+index_columns::copy_relative_offset_index() const noexcept {
+    return _relative_offset_index.copy();
+}
+chunked_vector<uint32_t>
+index_columns::copy_relative_time_index() const noexcept {
+    return _relative_time_index.copy();
+}
+chunked_vector<uint64_t> index_columns::copy_position_index() const noexcept {
+    return _position_index.copy();
+}
+
+void index_columns::assign_relative_offset_index(
+  chunked_vector<uint32_t> xs) noexcept {
+    _relative_offset_index = std::move(xs);
+}
+void index_columns::assign_relative_time_index(
+  chunked_vector<uint32_t> xs) noexcept {
+    _relative_time_index = std::move(xs);
+}
+void index_columns::assign_position_index(
+  chunked_vector<uint64_t> xs) noexcept {
+    _position_index = std::move(xs);
+}
+
+void index_columns::add_entry(
+  uint32_t relative_offset, uint32_t relative_time, uint64_t pos) {
+    _relative_offset_index.push_back(relative_offset);
+    _relative_time_index.push_back(relative_time);
+    _position_index.push_back(pos);
+}
+
+void index_columns::shrink_to_fit() {
+    vassert(
+      _relative_offset_index.size() == _relative_time_index.size()
+        && _relative_offset_index.size() == _position_index.size(),
+      "ALL indexes must match in size. {}",
+      *this);
+    _relative_offset_index.shrink_to_fit();
+    _relative_time_index.shrink_to_fit();
+    _position_index.shrink_to_fit();
+}
+
+index_columns index_columns::copy() const {
+    vassert(
+      _relative_offset_index.size() == _relative_time_index.size()
+        && _relative_offset_index.size() == _position_index.size(),
+      "ALL indexes must match in size. {}",
+      *this);
+    index_columns c;
+    c.assign_relative_offset_index(_relative_offset_index.copy());
+    c.assign_relative_time_index(_relative_time_index.copy());
+    c.assign_position_index(_position_index.copy());
+    return c;
+}
+
+bool index_columns::empty() const noexcept {
+    return _relative_offset_index.empty();
```

Comment:
Should you add the same `vassert()` you have in other functions on sizes here?

Reply:
done